nautobot/ui/src/js/theme.js

```diff
@@ -21,55 +15,84 @@ const getPreferredColorScheme = () =>
 const isValidTheme = (theme) => theme === THEME_DARK || theme === THEME_LIGHT || theme === THEME_SYSTEM;
 
 /**
- * Automatically detect Nautobot theme. It is derived from `cookie` or `localStorage` if set manually, and falls back to
- * preferred system color scheme by default.
- * @returns {('dark'|'light'|'system')} Detected Nautobot theme.
+ * Get preferred system color scheme from browser/OS settings.
+ * @returns {('dark'|'light')} Preferred system color scheme.
  */
-const detectTheme = () => {
-  const cookieTheme = getCookie('theme');
-  if (isValidTheme(cookieTheme)) {
-    return cookieTheme;
+const getPreferredColorScheme = () =>
+  window.matchMedia?.(`(prefers-color-scheme: ${THEME_DARK})`).matches ? THEME_DARK : THEME_LIGHT;
+
+/**
+ * Get the user's theme choice from localStorage, defaulting to 'system' if not set or invalid.
+ * @returns {('dark'|'light'|'system')} The user's theme choice.
+ */
+const getThemeChoice = () => {
+  let current_theme_choice = window.localStorage?.getItem('theme');
+  if (!isValidTheme(current_theme_choice)) {
+    current_theme_choice = THEME_SYSTEM;
   }
+  return current_theme_choice;
+};
+/**
+ * Determine the effective theme to be used.
+ * @returns {('dark'|'light')} The effective theme.
+ */
+const determineTheme = () => {
+  const current_theme_choice = getThemeChoice();
+  let determined_theme = getPreferredColorScheme();
 
-  const localStorageTheme = window.localStorage?.getItem('theme');
-  if (isValidTheme(localStorageTheme)) {
-    return localStorageTheme;
+  // If localStorage theme is valid and not 'system', we use the localStorage theme.
+  if (current_theme_choice !== 'system') {
+    determined_theme = current_theme_choice;
   }
 
-  return THEME_SYSTEM;
+  // Persist the determined theme in cookie for server-side access.
+  setCookie('theme', determined_theme);
+  return determined_theme;
+};
+
+/**
+ * Persist the user's theme choice in localStorage.
+ * @param {('dark'|'light'|'system')} theme - The user's theme choice.
+ * @returns {('dark'|'light'|'system')} The persisted theme choice.
+ */
+const persistThemeChoice = (theme) => {
```

Comment:
Feels confusing to me that `determineTheme()` sets the cookie, while `persistThemeChoice()` sets local storage. Shouldn't `persistThemeChoice` manage both?